### PR TITLE
Add static index signature

### LINKED ISSUE #6480: Allow defining static index signature
``` ts
class Foo {
    [key: string]: any;
    static BAR:string = 'bar';
    static BAZ:string = 'baz';
}

console.log(Foo['BAR']);
console.log(Foo['B' + 'A' + 'Z']);
console.log(Foo['qux']);
```

I get the following errors with this:

```
tst.ts(8,13): error TS7017: Index signature of object type implicitly has an 'any' type.
tst.ts(9,13): error TS7017: Index signature of object type implicitly has an 'any' type.
```

I've read in the gitter archives that this should be working. Should it? How can I get a static property value using a dynamic key?

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #6480

### CODE REVIEW

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by orta** [comment]
@typescript-bot pack this.

**Comment by sandersn** [comment]
@weswigham the reviewer list is broken on this PR, but you should look at this too

**Comment by Kingwl** [comment]
⬆️

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9664,6 +9677,13 @@ namespace ts {
             }
         }
 
+        function getIndexInfoOfIndexSymbol(indexSymbol: Symbol, indexKind: IndexKind) {
```
Comment (elibarzilay) [inline src/compiler/checker.ts:9680]: Since this is used only once in the above changed code, maybe make it a local helper instead?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -11094,24 +11114,29 @@ namespace ts {
         }
 
         function getIndexSymbol(symbol: Symbol): Symbol | undefined {
```
Comment (elibarzilay) [inline src/compiler/checker.ts:11116]: You're splitting both `getIndexSymbol` and `getIndexDeclarationOfSymbol` to two functions each, because your code calls them with different inputs, and it's the only place that needs the two new entry points.  One such refactor seems questionable, but two looks like too much.  (But OTOH I can't figure out if there's a way to avoid it.)

**Comment by elibarzilay** [comment]
Ping @DanielRosenwasser, @weswigham.

File: src/compiler/checker.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/checker.ts:9680]: I kinda prefer it this way to avoid several closure allocations, but it's likely a premature optimization.

Comment (elibarzilay) [reply src/compiler/checker.ts:9680]: (I'd be surprised if at least the v8 compiler wouldn't optimize it away...)

File: tests/baselines/reference/staticAsIdentifier.errors.txt [hunk 3]
```diff
@@ -1,10 +1,10 @@
-tests/cases/compiler/staticAsIdentifier.ts(2,5): error TS1071: 'static' modifier cannot appear on an index signature.
+tests/cases/compiler/staticAsIdentifier.ts(2,12): error TS1030: 'static' modifier already seen.
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply tests/baselines/reference/staticAsIdentifier.errors.txt:1]: No, I think it's worth discussing whether this is a regression. Shipping runtimes already accept

```js
class Bar {
  static static
}
```

[the remark this answers is not in the document]
Comment (elibarzilay) [reply tests/baselines/reference/staticAsIdentifier.errors.txt:1]: The regression point is there, but there are several existing tests that already have `static static`:
* `cases/compiler/multipleClassPropertyModifiersErrors.ts`
* `cases/compiler/staticModifierAlreadySeen.ts`
* `cases/conformance/parser/ecmascript5/IndexMemberDeclarations/parserIndexMemberDeclaration10.ts`
* `cases/conformance/parser/ecmascript5/MemberVariableDeclarations/parserMemberVariableDeclaration2.ts`

[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply tests/baselines/reference/staticAsIdentifier.errors.txt:1]: There is a difference in that `static static foo` on the same line doesn't exist today in ECMAScript.

Because there's a line terminator in this test, `static static` should continue to be whatever it is in ECMAScript due to automatic semicolon insertion rules.

In fact, I noticed that there's a bigger problem here which is that we currently error on code like

```ts
class C {
    static static
    foo() {}
}
```

which is actually totally acceptable JS. As a result, I've opened https://github.com/microsoft/TypeScript/issues/38283

**Comment by Kingwl** [comment]
Hi, Folks, I don't know what I can do. Should I work with the parser issue inside this PR?

**Comment by DanielRosenwasser** [comment]
I would say no - though, if you want to tackle the problem, coordinate with @rbuckton on #38283.

Otherwise, I'm not familiar enough with the internals. @elibarzilay and @weswigham, can you give this another review?

**Comment by Kingwl** [comment]
Seems something wrong if extends others

**Comment by Kingwl** [comment]
Seems eslint is broken, not my fault.

**Comment by Kingwl** [comment]
@typescript-bot test this.

**Comment by Kingwl** [comment]
@typescript-bot run tests.

File: tests/baselines/reference/staticAsIdentifier.errors.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (elibarzilay) [reply tests/baselines/reference/staticAsIdentifier.errors.txt:1]: @DanielRosenwasser, so this comment can be resolved given the separate issue?

File: src/harness/vfsUtil.ts [hunk 4]
```diff
@@ -663,7 +663,7 @@ namespace vfs {
             if (!isFile(node)) throw createIOError("EBADF");
 
             const buffer = this._getBuffer(node).slice();
-            return encoding ? buffer.toString(encoding) : buffer;
+            return encoding ? buffer.toString(encoding as any) : buffer;
```
Comment (elibarzilay) [inline src/harness/vfsUtil.ts:666]: Why is this needed?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -3645,6 +3645,19 @@ namespace ts {
...
 
+        function getNamedOrIndexSignatureMembers(members: SymbolTable): Symbol[] {
+            const result = getNamedMembers(members);
+            const index = getIndexSymbolFromSymbolTable(members);
+            if (!index) {
+                return result;
+            }
+            if (!result.length) {
+                return [index];
+            }
```
Comment (elibarzilay) [inline src/compiler/checker.ts:3656]: If this is done to avoid mutating `emptyArray` then maybe it's clearer to test `result === emptyArray`?  (Or if not, then a comment?)

**Comment by Kingwl** [comment]
up↑

**Comment by Kingwl** [comment]
Friendly ping. Up again.

**Comment by Kingwl** [comment]
Hey folks, Could this one be included into 4.0?

File: tests/cases/conformance/classes/staticIndexSignature/staticIndexSignature5.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// @strict: true
+
+class B {
+    static readonly [s: string]: number;
+    static readonly [s: number]: 42 | 233
+}
+
+interface I {
+    static readonly [s: string]: number;
```
Comment (weswigham) [inline tests/cases/conformance/classes/staticIndexSignature/staticIndexSignature5.ts:9]: Wait... What? Indexes don't have static sides (they describe a single shape) - shouldn't this be an error?

**Comment by Kingwl** [comment]
Up...

**Comment by Kingwl** [comment]
Uhhhhh... Up.

**Comment by weswigham** [comment]
@elibarzilay you wanna make sure this gets merged for 4.2? (After we cut the 4.1 RC in two weeks)

**Comment by Kingwl** [comment]
Emmmmm... Up

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by Kingwl** [comment]
As #41506.
@typescript-bot pack this.

**Comment by DanielRosenwasser** [comment]
It was mentioned in the last design meeting https://github.com/microsoft/TypeScript/issues/41538 that that this doesn't enforce anything on the property types, which is currently a blocker for the PR.

**Comment by Kingwl** [comment]
I'm sorry I'm not understand.

Would you mind provide an example for the case?

**Comment by weswigham** [comment]
```
class X {
    static [index: string]: string;
    static x = 12; // Should error, incompatible with index signature
}
class Y {
    static [index: string]: string;
    static foo() {} // should error, incompatible with index signature
}
```

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by sandersn** [comment]
@elibarzilay @weswigham I'd like to get this in early in the 4.3 cycle. Does this (1) have enough tests now (2) error correctly on property types?

**Comment by Kingwl** [comment]
Friendly ping 👀

**Comment by Kingwl** [comment]
👀PiNg AgAiN

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -3883,6 +3883,19 @@ namespace ts {
...
+        function getNamedOrIndexSignatureMembers(members: SymbolTable): Symbol[] {
+            const result = getNamedMembers(members);
+            const index = getIndexSymbolFromSymbolTable(members);
+            if (!index) {
+                return result;
+            }
+            if (result === emptyArray) {
+                return [index];
+            }
+            result.push(index);
```
Comment (weswigham) [inline src/compiler/checker.ts:3895]: Is this... Safe? Is `getNamedMembers` guaranteed to return a new array every call that it doesn't return `emptyArray`? Does it not return a cached result? If it returns a unique array every call, this is fine (maybe a comment noting the required invariant would be useful). If not, we probably would need to guarantee a unique, new result array here.

**Review by weswigham — APPROVED** [review]
Assuming my question is answered, I think this is probably good now, with @sandersn 's questions both answering in the affirmative. @elibarzilay since you're the PR assignee, you can probably merge this for this cycle whenever you're ready.

**Comment by Kingwl** [comment]
just for verify

@typescript-bot pack this.

**Comment by orta** [comment]
Hrm, I'm not certain why the playground didn't build (as this seems to be completely up to date with master), I'll look into add some more logs in the build process.

**Comment by Kingwl** [comment]
Try again.
@typescript-bot pack this.

**Review by DanielRosenwasser — APPROVED** [review]
